### ISSUE TITLE
research: JAX compile time — jit boundaries vs settings

### ISSUE THREAD

**Issue by Jammy2211** [PyAutoLabs/autolens_profiling#71]
## Overview

For complex likelihood functions, XLA compile time is now prohibitive — in the worst measured cases it exceeds the sampling time itself (A100 pixelized multi-start fit: ~35 min wall, almost entirely compile; 453 s to compute latent variables for a single sample; one `input_reduce_fusion` recompiling for >30 minutes). JAX-native samplers (`multi_start_adam`, the `searches_minimal` family) appear to make it worse than jit-only samplers, suggesting `jax.grad`/`value_and_grad` compounds the cost. This research task measures where compile time actually goes and delivers a verdict on the core question below.

**Core question:** do we need jit boundaries inside the source code to break up compilation into pieces, or can smaller source changes / JAX settings (persistent compilation cache, closure-cache fixes, compiler flags, tiling choices) get us there? The answer decides whether the follow-up is a config rollout or a library-wide restructuring of how JAX is threaded through the source.

## Plan

- Build a minimal compile-time probe in `autolens_profiling` (compile-vs-run split per case; reuse the existing artifacts/README pattern — no full dashboard, that is the companion feature task).
- Decompose compile cost along three axes: likelihood complexity (MGE → pixelized kernel-CDF → interferometer), JAX transform (`jit` vs `grad`/`value_and_grad` vs `vmap` vs `lax.map(batch_size=)`), and sampler structure (jit-only samplers vs JAX-native `multi_start_adam`).
- Test the cheap levers first: persistent compilation cache (`jax_compilation_cache_dir`), HLO dump (`--xla_dump_to=`) of the pathological `input_reduce_fusion`, `remat`/tiling alternatives to `lax.map`'s scan-of-vmap.
- Answer the core question directly: prototype one heavy likelihood split into piecewise jit boundaries and measure whether compilation breaks up profitably vs the same win from settings alone.
- Deliverable: a research note in `autolens_profiling` with before/after numbers and a verdict (settings suffice vs source restructuring needed); follow-up prompts filed via intake as their predecessors resolve.

<details>
<summary>Detailed implementation plan</summary>

### Affected Repositories
- autolens_profiling (primary — probe scripts + research note land here)
- autolens_workspace_developer (read-only evidence: `searches_minimal/*` scripts; repo is claimed by pixelized-gradient-experiment and is NOT claimed by this task)

### Branch Survey
| Repository | Current Branch | Dirty? |
|-----------|---------------|--------|
| ./autolens_profiling | main | clean |
| ./autolens_workspace_developer | main | dirty (pixelized-gradient-experiment holds the claim; read-only here) |

**Suggested branch:** `feature/jax-compile-time-research`

### Implementation Steps
1. Create `autolens_profiling/jax_compile/` probe module: time `jit`-compile vs first-eval vs steady-state eval per case, via `jax.block_until_ready` + `.lower()/.compile()` AOT split so compile seconds are measured directly rather than inferred.
2. Case matrix (small, curated — not the full cell grid): likelihood ∈ {MGE, pixelized kernel-CDF `RectangularKernelAdaptDensity(shape=(30,30), bandwidth=0.1, os_pix=1)`, interferometer}; transform ∈ {jit, value_and_grad, vmap, lax.map(batch_size=4)}; hardware: laptop RTX 2060 first (XLA_PYTHON_CLIENT_MEM_FRACTION=0.5), A100 via RAL `hpc/sync push-submit gpu` for the pathological cases only.
3. Levers, in cost order: (a) `jax_compilation_cache_dir` persistent cache — measure cold vs warm and cache-hit sensitivity to shape/closure changes; (b) `--xla_dump_to=` HLO dump of `input_reduce_fusion` on the pixelized value_and_grad case — identify what fuses into the f64[16, 15361, 2, 31, 512] monster; (c) `remat`/different tiling vs `lax.map` scan-of-vmap; (d) piecewise-jit prototype: split the pixelized likelihood at natural seams (tracing / mapping-matrix / solve) and measure total compile + runtime vs monolithic jit.
4. Research note `autolens_profiling/jax_compile/README.md` (or findings.md): numbers, verdict on the core question, and the spawned follow-ups.

### Key Files
- `autolens_profiling/searches/multi_start_adam/` — existing JAX-native sampler profiling case (worst-case evidence)
- `autolens_workspace_developer/searches_minimal/{multi_start_adam,gpu_multi_start_adam,optax_adam}.py` — sampler-structure reference (read-only)
- `autolens_workspace_developer/jax_profiling/gradient/` — gradient-probe pattern to reuse for compile timing

### Established — do not re-derive
- Autotuning is RULED OUT (2026-07-15: `--xla_gpu_autotune_level=0` vs on — 2090 s vs 2100 s, identical logL to the decimal).
- The same fusion compiled 7m36s under plain `value_and_grad` vs >30 min under `lax.map` scan-of-vmap — the sampler batching structure matters.
- Fresh-closure-per-call JIT cache-busting is a known stack trap; cache (closure, solver) on the instance.
- `analysis.print_vram_use()` triggers a full vmapped compile — not a cheap diagnostic on heavy cells.

### Companion task (separate, not duplicated here)
- `PyAutoMind draft/feature/autolens_profiling/jax_compile_time_profiling.md` — industrialize compile-time profiling across the full cell grid with regression tracking; consumes this task's findings and probe.

</details>

## Original Prompt

<details>
<summary>Click to expand starting prompt</summary>

JAX compile time is prohibitive for complex likelihood functions and JAX-native samplers.

We are now running into issues where for certain likelihood functions the time compiling them into JAX is prohibitive,
with a number of good examples in autolens_profiling.

In particular, this seems to have gotten even worse when we use JAX native samplers, like autolens_profiling/searches/multi_start_adam
and others in autolens_workspace_developer/searches_minimal (there are many). Maybe its jax.grad that is also compounding
the problem, with many samplers only use jax.jit.

This will become prohibitive for users, in some cases the compile time is longer than the sampling time.

Its time to research and asses what we can do about it:

**Core question (user clarification, 2026-07-16):** Do we need to put JAX compiles in the source code to break up compilation (e.g. jit boundaries inside the likelihood so it compiles in pieces), or can we speed things up via smaller source code changes or JAX settings (persistent compilation cache, donate/static argnums, avoiding closure cache-busts, compiler flags)? We may need to think hard about how JAX is implemented throughout the source code.

</details>

**Comment by Jammy2211** [comment]
**Parked (blocked): repo claim conflict.** `autolens_profiling` is currently claimed by task `slam-resume-profiling` (#70, branch `feature/slam-resume-profiling`, live session). Per the worktree claim contract this task is registered in `PyAutoMind/planned.md` and will resume straight into `start_workspace` (worktree `~/Code/PyAutoLabs-wt/jax-compile-time-research`, branch `feature/jax-compile-time-research`) once #70 frees the claim. Plan is human-approved (--auto supervised); no input needed unless you want this to jump the queue.

**Comment by Jammy2211** [comment]
**Unparked by human decision:** the two autolens_profiling tasks touch different parts of the repo (#70 owns the SLaM/resume paths; this task adds a new `jax_compile/` directory only). Parallel claim recorded in `active.md`. Worktree `~/Code/PyAutoLabs-wt/jax-compile-time-research` created on `feature/jax-compile-time-research`; research work starting now.

**Comment by Jammy2211** [comment]
**First measurement (probe committed, `81d28cf`):** MGE HST likelihood, CPU, AOT split:

| transform | trace | XLA compile | steady eval |
|---|---|---|---|
| `jit` | 29.0 s | 18.0 s | 0.12 s |
| `value_and_grad` | 30.0 s | **269.8 s** | 0.51 s |

`value_and_grad` multiplies XLA compile time **15×** on the *simplest* likelihood — XLA's slow-compile alarm fires on CPU. Tracing is transform-independent (~30 s, itself non-trivial). The 'jax.grad compounds it' hypothesis is confirmed at the first data point.

Running now (sequential, CPU): full MGE transform matrix (grad / vmap / vmap∘vag / lax.map∘vag batch_size=4 — the MultiStartAdam production shape), persistent-cache cold/warm pair, then pixelization jit+vag.

**Comment by Jammy2211** [comment]
**Full MGE CPU transform matrix** (resumed session; matrix had completed but was uncommitted — now `498fd0f`):

| transform | trace | XLA compile | steady eval |
|---|---|---|---|
| `jit` | 29.0 s | 18.0 s | 0.12 s |
| `grad` | 50.1 s | 231.8 s | 0.37 s |
| `value_and_grad` | 30.0 s | 269.8 s | 0.51 s |
| `vmap` | 12.0 s | 21.2 s | 2.38 s |
| `vmap∘vag` (n=16) | 49.7 s | 388.5 s | 25.2 s |
| `lax.map∘vag` (bs=4) | 125.5 s | **851.0 s** | 16.8 s |

Findings so far:
1. **Differentiation is the compile multiplier**: grad/vag ≈ 13–15× jit's XLA compile. Batching alone is nearly free (vmap ≈ 1.2× jit).
2. **`lax.map` multiplies it again**: 2.2× over vmap∘vag, 47× over jit, and 4× the trace time — consistent with the A100 observation (7m36s plain vag vs >30 min under lax.map, same fusion). The MultiStartAdam production shape pays both penalties.

Next (this session): persistent-cache cold/warm pair on `vag` (is the 270 s a one-time cost per machine?), then pixelization `jit`+`vag`, then the piecewise-jit prototype for the core question.

**Comment by Jammy2211** [comment]
**Persistent compilation cache: the settings lever works** (`jax_compilation_cache_dir`, cold/warm pair, MGE/vag/CPU, commit pending push):

| run | trace | XLA compile | steady |
|---|---|---|---|
| cache-cold | 17.1 s | 117.0 s | 0.20 s |
| cache-warm | 13.6 s | **2.3 s** | 0.16 s |

- **51× compile reduction warm**; the cache serves the AOT `.lower().compile()` path across processes. Residual warm cost is Python **tracing**, which no cache can remove (per-process) — tolerable at MGE scale (~14 s) but 125 s for `lax.map∘vag`.
- Production-scale corroboration from RAL job 330513 (A100 pixelized Nautilus): the **1h10m** `input_reduce_fusion` compile (~76 % of the job's 5518 s wall) **did serialize into the cache** (1.7 MB `jit_call` entry written at 16:19, the moment the alarm cleared).
- Caveat: earlier matrix numbers were load-inflated (vag 270 s vs 117 s idle) — the morning matrix ran alongside other work; relative ratios stand, absolute values need the idle-machine rule.

Next: (a) A100 warm-repeat of the 330513 shape — if the 1h10m vanishes, the cache is certified at the pathological scale; (b) pixelization jit+vag CPU probe; (c) piecewise-jit prototype for the remaining core-question half: can source-level jit boundaries cut the **cold** compile (first-fit UX) that the cache can't help.

**Comment by Jammy2211** [comment]
**CORRECTION to the transform matrix + new pyloop experiment** (idle machine, back-to-back, commit `c652455`):

| transform | trace | XLA compile | steady (full 16-start sweep) |
|---|---|---|---|
| `vmap∘vag` | 17.2 s | 124.2 s | 4.40 s |
| `lax.map∘vag` (bs=4) | 15.0 s | 116.8 s | 3.56 s |
| `pyloop_vag` (jit(vmap₄∘vag)×4 Python calls) | 14.7 s | 105.0 s | 3.01 s |

1. **The morning matrix's 388 s / 851 s were load artifacts** — the idle-machine rule strikes again. On idle CPU, every batched-gradient variant compiles in ~105–125 s, i.e. the same as plain `vag` (117 s idle). The honest CPU multipliers: differentiation ≈ **6×** jit (117 vs 18 s); batching structure (vmap vs lax.map vs Python loop) ≈ **free**.
2. **No CPU case for hoisting the batch boundary out of XLA**: `pyloop_vag` saves no compile and its sweep eval is merely competitive. The `lax.map` compile catastrophe is only evidenced in the **A100 logs** (7m36s vs >30 min, same fusion) — and those compiles ran on shared-node CPUs (8 cores/job), so even that comparison is load-suspect. → **Open item: controlled A100-node re-test** before the research note asserts a lax.map penalty.
3. Pixelization compiles FAST on CPU (jit 5.0 s / vag 30.7 s, sparse-operator config) — compile cost is **op-pattern-driven** (MGE's NNLS/linear-solve graph; GPU fusion heuristics), not proportional to model heaviness.

Standing findings: cache warm 117 s → 2.3 s (51×) unchanged; A100 warm-repeat (330534) in flight — already sampling at t≈2 min where the cold run compiled for 70 min.

**Comment by Jammy2211** [comment]
**A100 warm-repeat: cache certified at the pathological scale** (job 330534, exact arg-match re-run of 330513):

| run | wall | compile share |
|---|---|---|
| cold (330513) | 5517.8 s | ~4200 s (76 %) — one `input_reduce_fusion` |
| warm (330534) | **937.1 s** | none observed (sampling underway at t≈2 min) |

**5.9× end-to-end**, science consistent (max logL 17293 vs 17419, r_E 1.284 vs 1.310 — same basin, sampler stochasticity). The 1h10m compile is a **once-per-(machine, jax-version, shape)** cost when `jax_compilation_cache_dir` is set.

Verdict is firming up: **settings first, no restructure indicated so far.** Remaining evidence gaps before the research note: (1) controlled A100 lax.map-vs-plain-vag cold-compile comparison (the only remaining support for a structural penalty is load-suspect); (2) HLO dump of the pathological fusion (what IS being fused; upstream-XLA-report material); (3) whether piecewise source jit-boundaries could cut the **cold** 70 min for first-fit UX — only worth prototyping if (1)/(2) show the monolith itself, not the shared-CPU node, is the cause.

**Comment by Jammy2211** [comment]
## VERDICT: settings suffice — no source restructuring

Full research note: `jax_compile/README.md` on `feature/jax-compile-time-research`. Summary of the answer to the core question:

1. **Repeat fits — solved by one setting.** `jax_compilation_cache_dir`: 117 s → 2.3 s XLA compile locally (51×); A100 pixelized Nautilus end-to-end **5518 s → 937 s** (the 1h10m fusion compile vanishes; 1.7 MB cache entry). Follow-up filed: `PyAutoMind draft/feature/autofit/enable_the_jax_persistent_compilation_cache_by.md` (cache-by-default + first-compile UX log line).
2. **No structural defect to fix.** Controlled A/B on dedicated nodes: the pathological fusion compiles in 7m23s under `lax.map(vag, bs=4)` vs 7m24s under plain `vag` — identical. Idle-CPU: all batched-gradient structures compile in the same ~105–125 s band; `pyloop_vag` (batch boundary hoisted to Python) buys nothing. Historical ~15× penalties were host-load artifacts + the slow-compile alarm banner re-firing during ONE compile.
3. **Differentiation costs 11–15× compile over jit** (163.9/117.0 vs 10.9 s idle) — inherent to AD of the whole graph; the cache absorbs it.
4. **Piecewise source jit-boundaries: rejected by evidence** — the cost is one op-pattern fusion (kernel-CDF reduce on GPU; MGE linear-solve graph on CPU), not module size; splitting likelihoods would tax every eval to attack a once-per-machine cost the cache removes.
5. Compile cost is **op-pattern-driven, not complexity-driven**: pixelization (sparse-op config) compiles in 5 s/31 s on CPU — faster than parametric MGE.

Open artifact: HLO dump of the 7m30 fusion (job 330587 running) for an upstream XLA report — will attach to this issue when it lands. Measurement discipline note for future compile work: **XLA compiles on host CPUs — compile timings are load-sensitive even for GPU jobs**; idle machine or dedicated allocation only.

**Comment by Jammy2211** [comment]
**Ship gate checkpoint (supervised --auto):** research complete, branch ready (`feature/jax-compile-time-research`, 6 commits, additive `jax_compile/` only). Gate legs: tests **n/a** (no test dir in this workspace repo — stated), smoke **n/a** (new-dir additive, no downstream script surface), review **CLEAN**. Heart: **RED** — 5 reasons, all pre-existing and unrelated to this branch:

```
- PyAutoLens: 1 uncommitted source change(s)
- workspace validation not passing (3 failed, 2026-07-09T09-48-30Z)
- 58 stale parked script(s)
- manifest drift: tenant firewall (organ code) — 6 mismatch(es) vs PyAutoMind/repos.yaml
- release validation stale: source moved since rehearsal (PyAutoConf, PyAutoFit, PyAutoArray, PyAutoGalaxy, PyAutoLens)
```

Awaiting human ack to open the PR through unrelated RED (merge stays human regardless). HLO-dump artifact (job 330587) will be attached when it lands.

**Comment by Jammy2211** [comment]
**HLO artifact parked (open item):** `--xla_dump_to` is inert under jax 0.10.2 in this stack — zero files from an unfiltered A100 run (job 330587, flags confirmed in the job log, the 7m34s compile fired) and from a local CPU repro of the mechanism. The upstream XLA report needs its own small investigation (e.g. `jax.stages.Lowered.as_text()` on the lowered module); recorded in the research note. Not blocking: PR #73 is open with the verdict and awaits merge.

**Comment by Jammy2211** [comment]
**CORRECTION (found while implementing the cache rollout, PyAutoConf#127):** `--xla_dump_to` was never inert — `autoconf/jax_wrapper.py` **overwrites XLA_FLAGS at import**, silently discarding user/job flags. Fixed (append, not replace) on `feature/jax-cache-default`. Fallout:
- The HLO artifact just needs a re-run once the fix lands — un-parked.
- The historical 2026-07-15 'autotuning ruled out' A/B never flipped autotune (both runs clobbered to identical flags) — downgraded to **unproven**. The controlled lax.map-vs-vag A/B (jobs 330536/330537) is unaffected: both sides equally clobbered.
- Research note corrected on PR #73.

**Comment by Jammy2211** [comment]
Merged via PR #73 — research note + probe on main. Verdict: settings suffice (cache-by-default landed in PyAutoConf#128). Follow-up research prompt filed for cold-compile reduction (PyAutoMind draft/research/workspaces/investigate_ways_to_reduce_the_cold_jax.md). Closing.
